### PR TITLE
Add

### PR DESCRIPTION
**Pull request by yinlinchen**
- Batch atomic operations
- Versioning
- AuthZ
- Fixity

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -87,6 +87,78 @@ try (FcrepoResponse response = testClient.move(source, destination)
...
+try (FcrepoResponse response = testClient.post("fedoraurl/fcr:versions")
+        .slug("version1").perform()) {
+  URI destinationLocation = response.getLocation();
+}
+```
+
+Delete a version:
+```java
+try (FcrepoResponse response = testClient.delete("fedoraurl/fcr:versions/version1")
+      .perform()) {
```
Comment (awoods) [inline README.md]: Complete block with
```
   ...
}
```

File: README.md [hunk 2]
```diff
@@ -87,6 +87,78 @@ try (FcrepoResponse response = testClient.move(source, destination)
...
+```
+
+Revert a version:
+```java
+try (FcrepoResponse response = testClient.patch("fedoraurl/fcr:versions/version1")
+      .perform()) {
```
Comment (awoods) [inline README.md]: Complete block with
```
   ...
}
```

File: README.md [hunk 3]
```diff
@@ -87,6 +87,78 @@ try (FcrepoResponse response = testClient.move(source, destination)
...
+try (FcrepoResponse response = testClient.get("fedoraurl/tx:xxxx")
+        .perform()) {
+  String turtleContent = IOUtils.toString(response.getBody(), "UTF-8");
+}
+```
+
+Commit a transaction:
+```java
+try (FcrepoResponse response = testClient.post("fedoraurl/tx:xxxx/fcr:tx/fcr:commit")
+        .perform()) {
```
Comment (awoods) [inline README.md]: Complete block with
```
   ...
}
```

File: README.md [hunk 4]
```diff
@@ -87,6 +87,78 @@ try (FcrepoResponse response = testClient.move(source, destination)
...
+```
+
+Rollback a transaction:
+```java
+try (FcrepoResponse response = testClient.post("fedoraurl/tx:xxxx/fcr:tx/fcr:rollback")
+        .perform()) {
```
Comment (awoods) [inline README.md]: Complete block with
```
   ...
}
```

**Comment by awoods** [comment]
Let's take this ticket as an opportunity to make this README even more usable.
If users are incorporating this library into a Maven project, they need to know what dependency to include in their pom.xml:
```
    <dependency>
      <groupId>org.fcrepo.client</groupId>
      <artifactId>fcrepo-java-client</artifactId>
      <version>${fcrepo-client.version}</version>
    </dependency>
```
If wanting to use an log4j logger, include:
```
    <dependency>
      <groupId>ch.qos.logback</groupId>
      <artifactId>logback-classic</artifactId>
      <version>${logback.version}</version>
    </dependency>
```
Then we want to detail how to create a client:
```
// Create Fedora client
final FcrepoClient testClient = FcrepoClient.client().build();
```
Then we want to create a logical flow of actions: create resources, interact with resources, remove resources.
The README should be showing how to use the fluent patterns:
```
try (FcrepoResponse response = new PostBuilder(uri, testClient).body(text, "text/turtle").perform()) {
    location = response.getLocation();
    LOGGER.debug("Container creation status and location: {}, {}", response.getStatusCode(), location);
}
```

File: README.md [hunk 5]
```diff
@@ -87,6 +87,78 @@ try (FcrepoResponse response = testClient.move(source, destination)
 }
 ```
 
+Create a version:
+```java
+try (FcrepoResponse response = testClient.post("fedoraurl/fcr:versions")
```
Comment (awoods) [inline README.md]: These methods take `URI`s, not `String`s.

**Comment by acoburn** [comment]
This is a great start. And to @awoods' comments, I would add that, since Fedora is an LDP server implementation, it would be great if we encourage clients to interact with Fedora using the LDP interaction model. This would at least include an example of a client processing link headers (`rel="describedby"` and `rel="type"`).

**Comment by ajs6f** [comment]
I can't emphasize @acoburn 's point enough. We owe it to our users to model good behavior, and that means (amongst much else) showing them how to use the principle of HATEOS instead of URIs constructed by out-of-band rules.

**Comment by awoods** [comment]
@yinlinchen, It does not appear that you have addressed @acoburn's comment: https://github.com/fcrepo4-exts/fcrepo-java-client/pull/16#issuecomment-239579597
..which you may not be able to do completely.

# Versions
- `/fcr:versions` is only advertised after the first version is created on a resource
- Advertised via a triple on the resource with predicate `fedora:hasVersions`

# Fixity
- Should mention that this only applies to Binary resources
- Advertised on NonRdfSourceDescription via a triple with predicate `fedora:hasFixityService`

# Batch Atomic Operations
- Advertised on root resource via a triple with predicate `fedora:hasTransactionProvider`

# AuthZ
This section can be removed.

**Comment by awoods** [comment]
@acoburn, @ajs6f, regarding the question of advertising the services mentioned above, that is on the list of topics for the LDP-Next WG. I suspect it is best to leave the Fedora impl as it stands until a decision is made at the LDP level.

**Comment by yinlinchen** [comment]
@awoods If you can give a formal example on https://github.com/fcrepo4-exts/fcrepo-java-client/pull/16#issuecomment-239579597. I can update this PR. It is better for the author @acoburn to provide a complete example if you want to encourage other users to use this API.

**Comment by ajs6f** [comment]
@awoods `describedby` is part of the spec, [now.](https://www.w3.org/TR/ldp/#link-relation-describedby) To what question or process are you referring?

**Comment by yinlinchen** [comment]
@awoods BTW, see ticket description: https://jira.duraspace.org/secure/RapidBoard.jspa?rapidView=27&projectKey=FCREPO&view=detail&selectedIssue=FCREPO-2031 

Why AuthZ can be removed?

**Comment by awoods** [comment]
Without an example of [authorizationSparqlFile](https://github.com/fcrepo4-exts/fcrepo-java-client/pull/16/files#diff-04c6e90faac2675aa89e2176d2eec7d8R176) and [aclSparqlFile](https://github.com/fcrepo4-exts/fcrepo-java-client/pull/16/files#diff-04c6e90faac2675aa89e2176d2eec7d8R185), I am not sure there is any value in the [AuthZ](https://github.com/fcrepo4-exts/fcrepo-java-client/pull/16/files#diff-04c6e90faac2675aa89e2176d2eec7d8R172) section.

**Comment by awoods** [comment]
> @awoods describedby is part of the spec, now. To what question or process are you referring?

@ajs6f, I was referring to the question of how we can/should advertise `/fcr:versions`, `/fcr:fixity`, and `/fcr:tx` in a more LDP fashion.
Are you suggesting we somehow use `describedby`?

**Comment by ajs6f** [comment]
Those services are irrelevant. All @acoburn and I are asking for is a single example of using a header instead of a magic URI. `describedby` is enough for that.

**Comment by awoods** [comment]
"Those services" are relevant in the sense that it is helpful to give users of the fcrepo-java-client examples of how to use them.

**Comment by yinlinchen** [comment]
@awoods  See below, I am fine to remove AuthZ section.

```
authorizationSparqlFile:
PREFIX acl: <http://www.w3.org/ns/auth/acl#>
PREFIX pcdm: <http://pcdm.org/models#>

INSERT {
<> a acl:Authorization ;
acl:accessToClass pcdm:Object ;
acl:mode acl:Read, acl:Write;
acl:agent "adminuser" .
} WHERE { }
```
```
aclSparqlFile: 
PREFIX acl: <http://www.w3.org/ns/auth/acl#>

INSERT {
<> acl:accessControl </fcrepo/rest/test_authz/my-acls/acl>

} WHERE { }
```

**Comment by yinlinchen** [comment]
I will leave "describedby" example to you @awoods, @ajs6f , or @acoburn.  I am looking forward to seeing the best examples from you guys.  😁

**Comment by ajs6f** [comment]
No one is saying we shouldn't have examples of them. We _also_ need an example of proper use of header-based interaction.

**Comment by acoburn** [comment]
FWIW, there is a pointer to the transaction endpoint from the repo root (in the body of the RDF):

    <http://localhost:8080/fcrepo/rest/> fedora:hasTransactionProvider <http://localhost:8080/fcrepo/rest/fcr:tx>

So, while it's not a header to process, a client _could_ discover the location of that service endpoint.

**Comment by ajs6f** [comment]
Interesting. As a side note, should we open a ticket to change/augment that to also be provided in an LDP-style header?

**Comment by acoburn** [comment]
@ajs6f it would _exceptionally_ great if that "service URL" was available as a header, and ideally, it would be advertised on all resources, and not just the repo Root. It was my understanding that LDP-Next was going to discuss the issue of how to advertise these sorts of things.

As a concrete example, the camel machinery needs to know the "base URL" for the repo _a priori_ in order to efficiently use transactions (otherwise, how would it know to traverse up the container hierarchy?) If the TX service endpoint was available everywhere, it would make a lot of things much easier (for me, at least).

**Comment by ajs6f** [comment]
Well, there's a question as to whether this kind of link is the same as say, `fcr:versions` but either way, I think we're going to have to face the fact that we pretty much _are_ LDPNext.

**Comment by acoburn** [comment]
@ajs6f what I'd do is have this link header:

    Link: <http://repo.example.org/rest/fcr:tx>; rel="	http://fedora.info/definitions/v4/repository#hasTransactionProvider"

I'd expect the versions endpoint would have a different URI in the rel part. At least that's what other projects do (eg LDN)

**Comment by awoods** [comment]
@acoburn, that seems right.

**Comment by acoburn** [comment]
Except for the crummy formatting b/c I'm typing on a mobile device. But yes, you get the idea. The only problem w/ this approach is that you can get lots and lots of headers, which don't get compressed in HTTP/1.1. Another reason why HTTP/2 looks so very appealing.

**Comment by ajs6f** [comment]
Yeah, that seems pretty okay. Lots of headers doesn't bother me that much. Headers are metadata. I'd like eventually for the type of that link to be something that isn't owned directly by us, but that takes us right back to the LDPNext thing. :sigh:

**Comment by awoods** [comment]
@yinlinchen, I have created a PR against this current work:
https://github.com/yinlinchen/fcrepo-java-client/pull/1

If you merge the above PR with your branch, I will move the ticket forward:
https://jira.duraspace.org/browse/FCREPO-2031

**Pull request merged**